src/librustc_resolve/late.rs

```diff
@@ -217,7 +217,7 @@ impl<'a> PathSource<'a> {
             PathSource::Expr(parent) => match parent.map(|p| &p.kind) {
                 // "function" here means "anything callable" rather than `DefKind::Fn`,
                 // this is not precise but usually more helpful than just "value".
-                Some(&ExprKind::Call(..)) => "function",
+                Some(&ExprKind::Call(..)) => "function, tuple struct or tuple variant",
```

Comment:
Ugh, this is actually "any value that implements one of `Fn` traits", not just the listed things, the problem is that saying e.g. "callable value" here instead of "function" would look more cryptic.
On the other hand, the attempts to selectively list things here feel more wrong that just saying something generic.

I think we can improve things by using a heuristic here and say "function, tuple struct or tuple variant" only if the name starts with an uppercase letter, and say "function" otherwise.